evap/results/views.py

```diff
@@ -2,44 +2,74 @@
 
-    courses = Course.objects.filter(state__in=visible_states, semester__in=semesters).prefetch_related("degrees")
-    
-    courses = [course for course in courses if course.can_user_see_course(request.user)]
 
-    degrees = set()
-    course_types = set()
+def warm_up_template_cache(course):
+    assert course.state == 'published'
+    get_template('results_index_course.html').render(dict(course=course, can_user_see_results_page=True))
+    get_template('results_index_course.html').render(dict(course=course, can_user_see_results_page=False))
+
+
+def warm_up_template_cache_all():
+    courses = get_courses_with_prefetched_data(Course.objects.filter(state='published'))
     for course in courses:
-        if course.is_single_result:
+        warm_up_template_cache(course)
+
+
+def get_courses_with_prefetched_data(courses):
+    courses_with_prefetched_data = courses.prefetch_related(
+        "degrees",
+        "semester",
+        Prefetch("contributions", queryset=Contribution.objects.filter(responsible=True).select_related("contributor"), to_attr="responsible_contributions")
+    ).select_related("type")
+    for course in courses_with_prefetched_data:
+        course.responsible_contributors = [contribution.contributor for contribution in course.responsible_contributions]
+        if not course.is_single_result:
+            course.distribution = calculate_average_distribution(course)
+            course.avg_grade = distribution_to_grade(course.distribution)
+        else:
             course.single_result_rating_result = get_single_result_rating_result(course)
-        for degree in course.degrees.all():
-            degrees.add(degree)
-        course_types.add(course.type)
+    return courses_with_prefetched_data
 
-        course.distribution = calculate_average_distribution(course)
-        course.avg_grade = distribution_to_grade(course.distribution)
 
+@internal_required
+def index(request):
+    semesters = Semester.get_all_with_published_unarchived_results()
+    courses = Course.objects.filter(semester__in=semesters, state='published')
+    courses = [course for course in courses if course.can_user_see_course(request.user)]
 
+    if request.user.is_reviewer:
+        additional_courses = (Course.objects
+            .filter(semester__in=semesters, state__in=['in_evaluation', 'evaluated', 'reviewed'])
+            .annotate(num_participants=Count("participants", distinct=True), num_voters=Count("voters", distinct=True))
+        )
+        additional_courses = get_courses_with_prefetched_data(additional_courses)
+        courses += additional_courses
+
+    course_pks = [course.pk for course in courses]
+    degrees = Degree.objects.filter(courses__pk__in=course_pks).distinct()
+    course_types = CourseType.objects.filter(courses__pk__in=course_pks).distinct()
     template_data = dict(
         courses=courses,
-        degrees=sorted(degrees, key=lambda degree: degree.order),
+        degrees=degrees,
```

Comment:
the order is not guaranteed now anymore, right? it would be nice to make sure they are in the correct order when being added to the filter list.

Reply:
nevermind, we have the ordering defines in the model.